Bluetooth: Controller: BT_CTLR_ISO_TX_BUFFER_SIZE from BT_ISO_TX_MTU

cvinayak opened the pull request:
Derive BT_CTLR_ISO_TX_BUFFER_SIZE from BT_ISO_TX_MTU when BT_ISO is enabled.

Fixes #70794.

cvinayak committed:
Bluetooth: Controller: BT_CTLR_ISO_TX_BUFFER_SIZE from BT_ISO_TX_MTU

Derive BT_CTLR_ISO_TX_BUFFER_SIZE from BT_ISO_TX_MTU to have
optimal Controller memory allocations.
BT_CTLR_ISO_TX_BUFFER_SIZE can be set lower than
BT_ISO_TX_MTU in which case upper layer can send fragmented
SDU to the Controller.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on samples/bluetooth/broadcast_audio_source/overlay-bt_ll_sw_split.conf:
```diff
@@ -14,11 +14,6 @@ CONFIG_BT_CTLR_ADV_ISO_PDU_LEN_MAX=155
-# CONFIG_BT_ISO_TX_MTU value.
-#
-# Supports the highest SDU size required by any BAP LC3 presets (155)
+# Support the highest SDU size required by any BAP LC3 presets (155) + 8 bytes of HCI ISO Data
+# packet overhead (timestamp and HCI ISO Data packet header)
```
The core spec states for Read Buffer Size V2:

> Note: The LE_ACL_Data_Packet_Length and ISO_Data_Packet_Length return parameters do not include the length of the HCI ACL Data packet header or the HCI ISO Data packet header respectively.

So this should only need to be 155 + 4 = 159, as the HCI ISO Data packet header (4 octets) are _not_ included in the buffer size. 

As I understand it, all the places where you have `+ 8`, it should be `+ 4`, except the places where you use this to allocate memory for HCI ISO Data packets

cvinayak replied:
As mentioned in the comment in the file, 8 bytes is for the 4 bytes HCI ISO Data packet header plus 4 bytes for the optional timestamp (commented as the required overhead) if used by the upper layer. If only additional 4 bytes is allocated, and upper layer uses API with timestamp, then SDU would be fragmented over HCI.

jori-nordic replied:
yes @cvinayak and I got confused previously. @ivaniushkov fixed that miscalculation in the host a few months ago.

Thalley replied:
I'm still confused :D 

Is `CONFIG_BT_CTLR_ISO_TX_BUFFER_SIZE` the size of the buffers returned from the HCI command? If so, then it should _not_ include the 4 bytes header.

cvinayak replied:
i stand corrected... the 8 bytes is for the following:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/6350656/4d3597d9-0ace-480c-969f-944d72ed67fc)

Are we in sync? I will fix comments in this PR.

Thalley replied:
As I understand it, the TX buffer size from the controller is everything after the first 4 octets of Header (Conn handle + Data_Total_Length, etc.).

So for example to send an SDU of size 100 the buffer needs to be at least 100 + sizeof(psn) + sizeof(iso_sdu_length + status flag) = 104, or 108 if it include the timestamp. 

Today `CONFIG_BT_ISO_TX_MTU` determines the size of the SDU (i.e. the maximum value of ISO_SDU_length), but not the timestamp or the PSN, etc.  (that is added by the `BT_ISO_SDU_BUF_SIZE` macro). 

So if we want to derive `CONFIG_BT_CTLR_ISO_TX_BUFFER_SIZE` from `CONFIG_BT_ISO_TX_MTU` then we either need to:
1) `CONFIG_BT_CTLR_ISO_TX_BUFFER_SIZE = CONFIG_BT_ISO_TX_MTU + BT_HCI_ISO_TS_DATA_HDR_SIZE (8)`
2) Modify `BT_ISO_SDU_BUF_SIZE` so that it does not contain the timestamp and then do `CONFIG_BT_CTLR_ISO_TX_BUFFER_SIZE = CONFIG_BT_ISO_TX_MTU + BT_HCI_ISO_DATA_HDR_SIZE (4)`

So actually you were correct in using 8, but for the wrong reason.

cvinayak replied:
> So actually you were correct in using 8 (with timestamp), but for the wrong reason.

Agreed. @Emil-Gydesen-Bose  is the follow reasonably correct?
```
# Support the highest SDU size required by any BAP LC3 presets (155) + 8 bytes of HCI ISO Data
# packet overhead (the Packet_Sequence_Number, ISO_SDU_Length, Packet_Status_Flag fields; and
# the optional Time_Stamp field, if supplied)
```

cvinayak commented on subsys/bluetooth/controller/include/ll_feat.h:
```diff
@@ -189,10 +189,20 @@
 
+#if defined(CONFIG_BT_CTLR_ADV_ISO) || defined(CONFIG_BT_CTLR_CONN_ISO)
+#if defined(CONFIG_BT_ISO_TX_MTU)
+#define BT_CTLR_ISO_TX_BUFFER_SIZE MIN((CONFIG_BT_ISO_TX_MTU + 8U), \
+				       MAX(CONFIG_BT_CTLR_ISO_TX_BUFFER_SIZE, \
+					   8U))
```
Minimum value of `9U` will ensure a minimum SDU length of `1U`.

```suggestion
#define BT_CTLR_ISO_TX_BUFFER_SIZE MIN((CONFIG_BT_ISO_TX_MTU + 8U), \
				       MAX(CONFIG_BT_CTLR_ISO_TX_BUFFER_SIZE, \
					   9U))
```

Thalley commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -202,12 +202,12 @@ config BT_CTLR_ISO_TX_BUFFERS
 	depends on BT_CTLR_ADV_ISO || BT_CTLR_CONN_ISO
-	range 1 4095
-	default BT_ISO_TX_MTU if BT_ISO
+	range 1 65535
```
```suggestion
	range 1 16383
```

While the HCI command states that it has a range of 0x01 - 0xFFFF the `Data_Total_Length` is only 14 bits, so it doesn't really make sense to return/allocate buffers larger than that. 

In fact, since the ISO_SDU_Length is only 12 bits, the actual maximum buffer size is 4095 + 8 (timestamp, psn, length). 

We can keep it as 0xFFFF in case that the RFU bits will be used for the packet length in the future.

Thalley replied:
@cvinayak Will you follow through with this change? Or should it be 4095 + 8 or 16383?

cvinayak replied:
@Thalley i make the range be 9 to 4103 for `BT_LL_SW_SPLIT`  and left the HCI Spec range otherwise.

Thalley commented on subsys/bluetooth/controller/include/ll_feat.h:
```diff
@@ -189,10 +189,20 @@
 
+#if defined(CONFIG_BT_CTLR_ADV_ISO) || defined(CONFIG_BT_CTLR_CONN_ISO)
+#if defined(CONFIG_BT_ISO_TX_MTU)
+#define BT_CTLR_ISO_TX_BUFFER_SIZE MIN((CONFIG_BT_ISO_TX_MTU + 8U), \
```
Consider using the macro `BT_HCI_ISO_TS_DATA_HDR_SIZE` for places where you currently use `8` (or `BT_HCI_ISO_TS_DATA_HDR_SIZE + 1` instead of `9`).

cvinayak replied:
`struct bt_hci_iso_data_hdr`  naming is misleading, adds to the confusion!

Thalley replied:
Not fully disagree or agree with that :D But it's certainly better than magic numbers. 

But yeah, I see how the the data header can be confused with the packet header.

cvinayak committed:
Bluetooth: HCI: Rename to bt_hci_iso_sdu_hdr and bt_hci_iso_sdu_ts_hdr

Rename struct bt_hci_iso_data_hdr to bt_hci_iso_sdu_hdr, and
struct bt_hci_iso_ts_data_hdr to bt_hci_iso_sdu_ts_hdr.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on include/zephyr/bluetooth/hci_types.h:
```diff
@@ -108,17 +108,17 @@ struct bt_hci_acl_hdr {
 	uint16_t sn;
 	uint16_t slen; /* 12 bit len, 2 bit RFU, 2 bit packet status */
 } __packed;
-#define BT_HCI_ISO_DATA_HDR_SIZE	4
+#define BT_HCI_ISO_SDU_HDR_SIZE          4
 
-struct bt_hci_iso_ts_data_hdr {
+struct bt_hci_iso_sdu_ts_hdr {
 	uint32_t ts;
-	struct bt_hci_iso_data_hdr data;
+	struct bt_hci_iso_sdu_hdr sdu;
 } __packed;
-#define BT_HCI_ISO_TS_DATA_HDR_SIZE     8
+#define BT_HCI_ISO_SDU_TS_HDR_SIZE       8
 
```
I see what you are trying to do here, and I agree the naming could be improved. 

What I'm worried about here is that these are not headers for SDUs, but SDU fragments. 

If the host sends an SDU in 3 fragments to the controller, this header will appear 3 times. 

Would `bt_hci_iso_sdu_frag_xx` be better?

jori-nordic replied:
Will it though?
I think those "headers" are really part of the whole-sdu payload. Just like the SDU "header" in L2CAP K-frames, it shouldn't be repeated for every fragment. It's only there at the start of the SDU.

I would however be ok with renaming `struct bt_hci_iso_hdr` to `struct bt_hci_iso_data_hdr`:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/68428270/0d8b8e27-70ac-4d93-a0f5-dfa37214f088)

cvinayak replied:
> What I'm worried about here is that these are not headers for SDUs, but SDU fragments.

I disagree here, by the Spec:

```
If PB_Flag equals 0b00 or 0b10, then the Packet_Sequence_Number,
ISO_SDU_Length, and Packet_Status_Flag fields (plus the intermediate RFU
field) shall all be present in the packet and the Time_Stamp field may be
present. If PB_Flag equals 0b01 or 0b11, then none of these fields shall be
included in the packet.
```

Hence, `struct bt_hci_iso_sdu_hdr` is once per SDU (first fragment or complete SDU) from the upper layer, these fields are not part of next SDU fragment towards the Controller.

cvinayak replied:
> I would however be ok with renaming struct bt_hci_iso_hdr to struct bt_hci_iso_data_hdr

This I can do, and will be closer to the Spec wording `HCI ISO Data packet header`.

Thalley replied:
> > What I'm worried about here is that these are not headers for SDUs, but SDU fragments.
> 
> I disagree here, by the Spec:
> 
> ```
> If PB_Flag equals 0b00 or 0b10, then the Packet_Sequence_Number,
> ISO_SDU_Length, and Packet_Status_Flag fields (plus the intermediate RFU
> field) shall all be present in the packet and the Time_Stamp field may be
> present. If PB_Flag equals 0b01 or 0b11, then none of these fields shall be
> included in the packet.
> ```
> 
> Hence, `struct bt_hci_iso_sdu_hdr` is once per SDU (first fragment or complete SDU) from the upper layer, these fields are not part of next SDU fragment towards the Controller.

I stand corrected. Thanks! 

In this case, I'm OK with the new naming

jori-nordic replied:
super confusing as always from the SIG. And the next bit in the same paragraph is cool too "SDU fragments generated over HCI are unrelated to the SDU fragments generated by ISOAL".

Thalley replied:
Indeed. The entire ISO over HCI is a complete mess :(

Thalley commented on subsys/bluetooth/controller/include/ll_feat.h:
```diff
@@ -189,10 +189,22 @@
 	defined(CONFIG_BT_CTLR_PERIPHERAL_ISO)
 #define LL_CIS_OCTETS_TX_MAX MIN(CONFIG_BT_CTLR_CONN_ISO_PDU_LEN_MAX, \
-				 CONFIG_BT_CTLR_ISO_TX_BUFFER_SIZE)
+				 (BT_CTLR_ISO_TX_BUFFER_SIZE - 8U))
```
```suggestion
				 (BT_CTLR_ISO_TX_BUFFER_SIZE - BT_HCI_ISO_SDU_TS_HDR_SIZE))
```

Thalley commented on subsys/bluetooth/controller/include/ll_feat.h:
```diff
@@ -202,7 +214,7 @@
 #define LL_FEAT_BIT_ISO_BROADCASTER BIT64(BT_LE_FEAT_BIT_ISO_BROADCASTER)
 #define LL_BIS_OCTETS_TX_MAX MIN(CONFIG_BT_CTLR_ADV_ISO_PDU_LEN_MAX, \
-				 CONFIG_BT_CTLR_ISO_TX_BUFFER_SIZE)
+				 (BT_CTLR_ISO_TX_BUFFER_SIZE - 8U))
```
```suggestion
				 (BT_CTLR_ISO_TX_BUFFER_SIZE - BT_HCI_ISO_SDU_TS_HDR_SIZE))
```
